### DISCUSSION TITLE
INPUT WANTED: CSLA for netcore3.1 dependencies

### DISCUSSION THREAD

**Discussion by rockfordlhotka** [MarimerLLC/csla#1985] [General]
I'd like some thoughtful input from the community.

Right now CSLA 5.4.0, in the `Csla` NuGet package, includes a netstandard2.0 assembly and a net5.0 assembly.

If you reference CSLA in a net5 project, you'll get the net5 `Csla.dll`, otherwise you'll get the ns20 `Csla.dll` (or the .NET 4, 4.5, or 4.6 versions).

Here's my question: if you reference the .NET 4.6 or netstandard2.0 versions, would you expect your project to get updated to the net5 `Microsoft.*` dependencies CSLA requires? Or would you expect to only get the net5 dependencies if you are building a net5 project?

This is important, because it is absolutely possible for a netstandard2.0 project (like Csla) to reference net5 dependencies. In fact, the automated Dependabot in GitHub keeps pestering me to do exactly that, because they are "the most current" available.

So far I have been resisting that change, because I have this gut feeling that some of you are targeting netcore3.1, using the ns20 `Csla.dll` _and don't want any net5 anything in your projects_.

**Is that true???***

**Comment by BlagoCuljak** [comment]
We have NS2.0 and will try to upgrade to net5, some compatibility issues exist, and we are trying to find suitable replacement.

Change will come, sooner or later, and it's hard to say when will be all on same net5 dependency ship, but from my point of view, people are still waiting to get on net5, or even waiting LTS, net6 to upgrade. 

Why wouldn't someone want net5, latest and greatest, it's not clear to me.

**Reply by ajohnstone-ks** [reply]
> Why wouldn't someone want net5, latest and greatest, it's not clear to me.

Because Xamarin.Forms is built on Mono, not .Net 5, and Csla shouldn't pull in a different .net runtime.

**Comment by hurcane** [comment]
I'm going to respond with more questions because I haven't taken the time to learn how all these different versions of .NET affect our environment.

Our product using CSLA is primarily deployed as an on-premise installation via Windows Installer. The client targets .NET 4.7.2 and the ASP.NET Core project targets .NET Core 3.1.

The installer includes a bootstrap process that checks for prerequisite products that are not installed, such as the IIS hosting bundle for .NET Core.  If CSLA included the .NET 5 version of the dependencies, would this mean that the installer's bootstrap logic would need to check whether .NET 5 is installed?

**Reply by rockfordlhotka** [reply]
There are two separate things here.

First, there are various `Microsoft.*` packages that continue to support .NET 4.7.2+, netcore3.1, and now net5.0. Most of these packages have progressed from _their_ version 3.1.x to _their_ version 5.0. Not specifically having to do with the runtime they target.

Second, many of us (at least library authors) now need to support .NET 4.7.2 (or 4.8), netcore3.1, and net5.0. In many cases we can get the first two by targeting netstandard2.0, and now also net5.0.

`Csla.dll` in the `Csla` package currently builds for the following:

* .NET 4.0
* .NET 4.5
* .NET 4.6.1
* netstandard2.0 (used on xamarin, uwp, netapp3.1)
* net5.0

It has dependencies on numerous `Microsoft.*` packages. Mostly today at version 3.1.9 or 3.1.10.

The question is whether to allow them _all_ to use `Microsoft.*` packages at version 5.0 - regardless of which runtime they target.

And you know, as I type this I think the answer is _yes_. Yes they _all_ should target the latest `Microsoft.*` version, as long as that version supports all the existing runtimes.

**Reply by hurcane** [reply]
I cannot imagine a a better explanation! From my perspective, I would agree that targeting the latest dependency version that supports all the existing supported runtimes makes sense.

**Reply by BlagoCuljak** [reply]
We have net 4.7.2 WPF app, and our Business Library was NetStandard2.0, and it was compatible.

After updating Business Library and App Server to Net5, I get error in ViewModels Project:
`BusinessLibrary.Standard is not compatible with net472 (.NETFramework,Version=v4.7.2) / win-x64. Project BusinessLibrary.Standard supports: net5.0`

So, if we are targeting dependency later then 5.0., new versions of CSLA will not be able to be used on older net projects, even on 4.8.

**Reply by ajohnstone-ks** [reply]
I think Csla netstandard should only use netstandard.  

My understanding is that netstandard is not an implementation at all, but instead a set of contracts that various implementions must implement to be netstandard compliant.   Only the top level application building something that actually gets run pulls in the proper runtime to use which contains implementations of those.

I think if Csla depends on .net5 only things, it might cause issues for code running against .net framework or mono.  At least that's my understanding of things, I could be way off 😄
